src/swap/saga.ts

```diff
@@ -0,0 +1,116 @@
+import { CeloTx } from '@celo/connect'
+import { TxParamsNormalizer } from '@celo/connect/lib/utils/tx-params-normalizer'
+import { ContractKit } from '@celo/contractkit'
+import { PayloadAction } from '@reduxjs/toolkit'
+import { call, put, select, takeLatest } from 'redux-saga/effects'
+import { maxSwapSlippagePercentageSelector } from 'src/app/selectors'
+import { navigate } from 'src/navigator/NavigationService'
+import { Screens } from 'src/navigator/Screens'
+import {
+  swapApprove,
+  swapError,
+  swapExecute,
+  swapPriceChange,
+  swapStart,
+  swapSuccess,
+} from 'src/swap/slice'
+import { Field, SwapInfo, SwapTransaction } from 'src/swap/types'
+import { sendTransaction } from 'src/transactions/send'
+import { newTransactionContext } from 'src/transactions/types'
+import { fetchWithTimeout } from 'src/utils/fetchWithTimeout'
+import Logger from 'src/utils/Logger'
+import { getContractKit } from 'src/web3/contracts'
+import networkConfig from 'src/web3/networkConfig'
+import { getConnectedUnlockedAccount } from 'src/web3/saga'
+import { applyChainIdWorkaround, buildTxo } from 'src/web3/utils'
+
+const TAG = 'swap/saga'
+
+function getPriceDiff(price1: number, price2: number) {
+  return (Math.abs(price1 - price2) / ((price1 + price2) / 2)) * 100
+}
+
+export function* swapSubmitSaga(action: PayloadAction<SwapInfo>) {
+  try {
+    // Navigate to swap pending screen
+    yield call(navigate, Screens.SwapPending)
+
+    // Check that our guaranteedPrice is within 2%, maxSwapSlippagePercentage, of of the price
+    const maxSlippagePercent: number = yield select(maxSwapSlippagePercentageSelector)
+    const priceDiff: number = yield call(
+      getPriceDiff,
+      +action.payload.unvalidatedSwapTransaction.price,
+      +action.payload.unvalidatedSwapTransaction.guaranteedPrice
+    )
+    if (priceDiff >= maxSlippagePercent) {
+      yield put(swapPriceChange())
+      return
+    }
+
+    // Set contract kit, wallet address and normalizer
+    const kit: ContractKit = yield call(getContractKit)
+    const walletAddress: string = yield call(getConnectedUnlockedAccount)
+    const normalizer = new TxParamsNormalizer(kit.connection)
+
+    // Approve transaction
+    yield put(swapApprove())
+    Logger.debug(TAG, `Starting to swap approval for address: ${walletAddress}`)
+    const rawApproveTx = { ...action.payload.approveTransaction, from: walletAddress }
+    applyChainIdWorkaround(rawApproveTx, yield call([kit.connection, 'chainId']))
+    const approveTx: CeloTx = yield call(normalizer.populate.bind(normalizer), rawApproveTx)
+    const approveTxo = buildTxo(kit, approveTx)
+    yield call(
+      sendTransaction,
+      approveTxo,
+      walletAddress,
+      newTransactionContext(TAG, 'Swap/Approve')
+    )
```

Comment:
what do you think about abstracting this block, since we use it later below? i was thinking something like `yield call(sendSwapTransaction, action.payload.approveTransaction, newTransactionContext(TAG, 'Swap/Approve')` would do the trick? the benefit i see with this is being able to move all the contractkit logic away from the swap logic and making this saga more focused on the swap details. also not seeing `applyChainIdWorkaround` twice is nice haha

Reply:
I tried this in 607125137e11d2a0cbed357d7ddf1dfade6164e6, but ran into issues with the swaps executing. Maybe something to do with passing around the kit and normalizer?

Reply:
Ran into some issues with the chain Id. Can attempt in a follow up pr.

Reply:
i was playing around with this and it seemed to work for me - strangely i didn't really change anything from what we tried yesterday. since i already have it i'll submit the PR for you to have a look